### PR TITLE
Bug 1840332 - Create ReviewQualityCheckState

### PR DESCRIPTION
**Pull request by rahulsainani**
### How
– Creates UIState
– Makes `ReviewGrade` private to the composable as that contains info for the component specifically and the state includes `Grade`.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1840332

### CODE REVIEW

**Review by MozillaNoah — COMMENTED** [review]
this is looking good so far! I've left a couple questions and nits for this first pass.

My main points are:
- Could we get a high level of what you're envisioning with the state/store here? It looks like you're going for a state machine model of `State` and using a mega `when` block in the root Composable to control the UI?
- Historically, the action, store, state, and reducer are all in one file. I don't know why they were implemented this way, but maybe @jonalmeida could provide some more context on this part.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    enum class Grade {
+        A, B, C, D, F
+    }
+
+    /**
+     * Factors for which highlights are available based on recent reviews of the product.
+     */
+    enum class HighlightType {
+        QUALITY, PRICE, SHIPPING, PACKAGING_AND_APPEARANCE, COMPETITIVENESS
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt:93]: They are review highlights, I think the doc would basically just state what type of highlight it is. [Figma](https://www.figma.com/file/VE5xayz3yOSB5qOATUU9Im/Mobile-Shopping-Experience-MVP?type=design&node-id=3708-74326&mode=design&t=9GoRxdf4lAxg98Bj-0) for UIState. I'm not sure the doc adds value here, but I don't have a strong opinion here.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckStore.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+package org.mozilla.fenix.shopping.state
+
+import mozilla.components.lib.state.Store
+
+/**
+ * Store for review quality check feature.
+ */
+class ReviewQualityCheckStore : Store<ReviewQualityCheckState, ReviewQualityCheckAction>(
+    initialState = ReviewQualityCheckState.Initial,
+    reducer = { _, _ -> ReviewQualityCheckState.Initial },
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckStore.kt:14]: Yeah, will have iterations of functioning reducer in the next patches.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewGrade.kt [hunk 3]
```diff
@@ -172,6 +175,18 @@ private fun ReviewGradeLetter(
...
+ * Maps [Grade] to [ReviewGrade].
+ */
+private fun Grade.toReviewGrade(): ReviewGrade =
+    when (this) {
+        Grade.A -> ReviewGrade.A
+        Grade.B -> ReviewGrade.B
+        Grade.C -> ReviewGrade.C
+        Grade.D -> ReviewGrade.D
+        Grade.F -> ReviewGrade.F
+    }
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewGrade.kt:188]: What will be the benefits of moving there? 

There might be an extra accessor method overhead as that'll be a companion object function.

The benefit of this is `ReviewGrade` doesn't know about `Grade` and this just a mapper function, usage looks good.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt:93]: Gotcha gotcha gotcha. If they're more for pairing up the API response to sections of the UI and don't drive specific business logic beyond that, I don't think the docs are super necessary.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                val adjustedRating: Float,
+                val productUrl: String,
+                val highlights: Map<HighlightType, List<String>>?,
+                val recommendedProductState: RecommendedProductState = RecommendedProductState.Initial,
+            ) : ProductReviewState
+        }
+    }
+
+    /**
+     * Review Grade of the product - A being the best and F being the worst.
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt:83]: I wonder if it's worth noting 'E' is intentionally not a grade

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+             */
+            object Loading : ProductReviewState
+
+            /**
+             * Denotes an error has occurred.
+             */
+            object Error : ProductReviewState
+
+            /**
+             * Denotes no analysis is present for the product user is browsing.
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt:50]: nit:
```suggestion
             * Denotes no analysis is present for the product the user is browsing.
```

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+             *
+             * @property productUrl The url of the product user is browsing.
```
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckState.kt:52]: nit:
```suggestion
             * @property productUrl The url of the product the user is browsing.
```

**Review by t-p-white — APPROVED** [review]
Looking good, nice work 👍 
Couple of optional nits that I missed the first time

**Comment by rahulsainani** [comment]
>     * Could we get a high level of what you're envisioning with the state/store here? It looks like you're going for a state machine model of `State` and using a mega `when` block in the root Composable to control the UI?

In a way, yes. The idea is to model the state as [the designs](https://www.figma.com/file/VE5xayz3yOSB5qOATUU9Im/Mobile-Shopping-Experience-MVP?type=design&node-id=2678-62309&mode=design&t=F0SdJTSuIcWGSsjB-0). As we can see there is a opted out and opted in state at the root level. So one can think of the top level UI to have a when block with 2 options. Then the `OptedIn` UI can handle it's own states - loading, error, content, etc. I think this will  become clearer as we add the UI components and observe the state. 


>     * Historically, the action, store, state, and reducer are all in one file. I don't know why they were implemented this way, but maybe @jonalmeida could provide some more context on this part.

Happy to change it going forward  👍

**Comment by jonalmeida** [comment]
> Historically, the action, store, state, and reducer are all in one file. I don't know why they were implemented this way, but maybe jonalmeida could provide some more context on this part.

@rahulsainani @MozillaNoah this is fine. We keep each in separate files, but in Fenix with store-specific screens which are smaller stores, that wasn't really enforced.

**Pull request merged**